airflow/models/taskinstance.py

```diff
@@ -4036,6 +4036,12 @@ def clear_db_references(self, session: Session):
                 )
             )
 
+    def get_upstream_task_ids_by_state(self, state: list[TaskInstanceState], session: Session) -> set[str]:
+        """Get direct upstream tasks ids by state."""
+        task_instances = self.dag_run.get_task_instances(state=state, session=session)
+        dag_task_instances = set(task_instance.task_id for task_instance in task_instances)
+        return dag_task_instances & self.task.upstream_task_ids  # type: ignore[union-attr]
```

Comment:
It’s probbly better to do this the other way around? i.e. get the upstream task IDs first, and then filter them by state. The performance difference is likely significant if the DAG is pretty large (and `get_task_instances` may need to fetch a lot more rows than needed).